### PR TITLE
zz/jm - Add image

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [x] The landing page links to the image link submission form.
- [x] The link is entered through a form.
- [x] After the form is saved, the link is persisted in the database.
- [x] After the form is submitted, the user is redirected to a page displaying
  the image.
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.
- [x] A flash message is displayed indicating the image was created successfully

__Discussion Topic__:
- [x] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by zzjas**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [X] I have replaced `X` above with the associated issue number.
- [X] I have changed the issue's label to "pull request".
- [X] I have assigned the designated reviewer to this PR.
- [X] I have set the appropriate label on this PR.
- [X] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/assets/javascripts/images.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (jmangel) [inline app/assets/javascripts/images.js]: I think we can remove these automatically generated files that we don't use, `images.js`, `images.scss`, `images_helper.rb`

File: db/schema.rb [hunk 2]
```diff
@@ -10,6 +10,12 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 0) do
+ActiveRecord::Schema.define(version: 2019_06_18_174030) do
```
Comment (jmangel) [inline db/schema.rb]: There's a lot going on in this first commit, could we split this out so that we add the database table (the migration file) and the model (and test) in their own commit? This way we'll have one atomic commit that can stand on its own, has a clear purpose ("add Image model to database"), and builds the foundations for the next commits

File: db/schema.rb [hunk 3]
```diff
@@ -10,6 +10,12 @@
...
+
+  create_table "images", force: :cascade do |t|
+    t.string "url"
```
Comment (jmangel) [inline db/schema.rb]: Since we know we will never want to have an image in the database that doesn't have a url, let's forbid blank urls with `t.string "url", null: false`. It's a nice thing to thing about when adding new database fields

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -5,15 +5,29 @@ def new
 
   def create
     @image = Image.new(image_params)
-    @image.save
-    redirect_to new_image_path
+    if @image.valid?
```
Comment (jmangel) [inline app/controllers/images_controller.rb]: `@image.save` will return `false` if the validations fail, so we can probably skip this check, get rid of the `save_image` method, and do 
```
if @image.save
  flash[:notice] = 'Image was successfully created.'
  redirect_to @image
else
  render :new
end
```

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -5,15 +5,29 @@ def new
...
+      save_image
+    else
+      render :new
+    end
   end
 
-  def show; end
+  def show
+    @image = Image.find(params[:id])
```
Comment (jmangel) [inline app/controllers/images_controller.rb]: This line is added in the commit to `Add success&failure messages`, but it seems like it doesn't belong there. We can work together to reorganize those commits, but I think it'd be good to have one separate commit for `add image show page`, which needs to come before the commit that adds the functionality to redirect to the show page when the image is created

File: app/views/images/new.html.erb [hunk 6]
```diff
@@ -3,4 +3,8 @@
 <%= form_for @image do |f| %>
     <%= f.text_field :url %>
     <%= f.submit "Create" %>
+<% end %>
+
+<% if @invalid_image_url %>
```
Comment (jmangel) [inline app/views/images/new.html.erb]: Since we ended up using `simple_form_for` instead of `form_for` and this custom error message, we probably don't need to commit this intermediate stage. With rebasing, this is not so hard. I can work on that with you

File: app/models/image.rb [hunk 7]
```diff
@@ -1,2 +1,3 @@
 class Image < ApplicationRecord
+  validates :url, presence: true
```
Comment (jmangel) [inline app/models/image.rb]: This url validation (and test) probably belong in their own commit, or at least in the same commit I mentioned to add the model and migration

File: Gemfile [hunk 8]
```diff
@@ -19,6 +19,9 @@ gem 'jquery-rails'
 
 gem 'webpacker', '~> 4'
 
+gem 'simple_form'
```
Comment (jmangel) [inline Gemfile]: This commit is called `Add image validator and tests`, so it probably shouldn't include the addition of `simple_form`. We can create a commit to "add new image form" and put the addition of `simple_form` gem in there

File: app/controllers/images_controller.rb [hunk 9]
```diff
@@ -1,13 +1,15 @@
 class ImagesController < ApplicationController
   def new
     @image = Image.new
+    @not_new_url = false
```
Comment (jmangel) [inline app/controllers/images_controller.rb]: It doesn't look like we need this variable anymore

File: lib/templates/erb/scaffold/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<%# frozen_string_literal: true %>
+<%%= simple_form_for(@<%= singular_table_name %>) do |f| %>
```
Comment (jmangel) [inline lib/templates/erb/scaffold/_form.html.erb]: What is this file needed for?

File: test/models/image_test.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  test 'validation for correct url' do
+    new_image_valid_should_be @test_url + '.png', true
+  end
+
+  test 'validation for empty' do
+    new_image_valid_should_be '', false
+  end
+
+  test 'validation for invalid url' do
+    new_image_valid_should_be @test_url, false
```
Comment (jmangel) [inline test/models/image_test.rb]: Can we also test the error message in these tests, like we do with `wrong_image_should_have_error_msg_of`?

File: test/models/image_test.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+  test 'validation for invalid format' do
+    new_image_valid_should_be @test_url + '.pdf', false
+    wrong_image_should_have_error_msg_of @test_url + '.pdf', 'Invalid file format'
+  end
+
+  private
+
+  def new_image_valid_should_be(url, expect)
+    assert_equal Image.new(url: url).valid?, expect
```
Comment (jmangel) [inline test/models/image_test.rb]: There is an assertion method we can use that has better error messages than `assert_equal` in the case where you're testing true or false. It's called `assert_predicate`/`refute_predicate`, and it's used like `assert_predicate Image.new(url: url), :valid?`. `:valid?` is a symbol representing the `.valid?` method, and `assert_predicate` asserts that it returns true, `refute_predicate` asserts that it returns false. So I don't think we need a helper method, we can probably just use that directly in our tests

File: Gemfile (hunk 8 above)
Comment (jmangel) [inline Gemfile]: It seems like we could split this commit further into "add Image model and table" and "add simple_form dependency"

File: db/schema.rb [hunk 13]
```diff
@@ -10,6 +10,12 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 0) do
+ActiveRecord::Schema.define(version: 2019_06_18_174030) do
+
+  create_table "images", force: :cascade do |t|
+    t.string "url", null: false
```
Comment (jmangel) [inline db/schema.rb]: Sorry, I should've left the comment about `null: false` in the migration file. Really, you should never change the schema directly, since it gets edited by the migration. The better way to do this would be to `rake db:rollback` to undo the migraiton, set `t.string :url, null: false` in the migration, and then re-run the migration, which will set the schema

File: test/models/image_test.rb [hunk 14]
```diff
@@ -11,19 +11,19 @@ class ImageTest < ActiveSupport::TestCase
 
   test 'validation for empty' do
     image = Image.new(url: '')
-    refute_predicate image, :valid?
-    assert image.errors.details[:url].any? { |e| e[:error] == :blank }
+    assert_not_predicate image, :valid?
```
Comment (jmangel) [inline test/models/image_test.rb]: I think these changes should go in the "add Image model and table" commit instead of "add image show page" commit

File: test/controllers/image/images_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    assert_response :success
+    form_exists ''
+  end
+
+  test 'should create image' do
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.png' } }
+    end
+
+    assert_redirected_to image_path(Image.last)
```
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: At this point in the commits (" Add new image page"), the image show page doesn't exist, so we probably shouldn't test it there. We could either edit this commit to have an intermediate version of the test, or (maybe easier), we could move the commits so that the show page is created before the new/create action. What do you think?

File: app/views/home/index.html.erb [hunk 16]
```diff
@@ -1 +1,2 @@
-<h1>Hello world!</h1>
\ No newline at end of file
+<h1>Hello world!</h1>
+<%= link_to 'Image Submission', new_image_path %>
```
Comment (jmangel) [inline app/views/home/index.html.erb]: Could we add a test for this in the controller test?

File: test/controllers/image/images_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    assert_equal 'Image was successfully created.', flash[:notice]
+  end
+
+  test 'shoud stay when fail to create new image' do
+    img = { image: { url: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.pdf' } }
+    assert_no_difference('Image.count') do
+      post images_path, params: img
+    end
+
+    form_exists img[:image][:url]
```
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: Here, we can also test that the response is a failure response with `assert_response :unprocessable_entity`

File: test/controllers/image/images_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
```
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: Can we add some tests for the `show` action? Minimally, we should test a successful case (renders the correct image) and a failure case (image does not exist, `get image_path(-1)`)

File: test/controllers/image/images_controller_test.rb (hunk 17 above)
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: What do you think about adding a test that the error shows up? I know that we decided not to test this, because it is a responsibility of `simple_form_for`, but if `simple_form_for` changed so it didn't show the error, then we would probably want to see our test fail so that we could at least add a generic flash message like "image could not be saved".

Comment (zzjas) [reply test/controllers/image/images_controller_test.rb]: I added one test for simple_form just to see if it prints something. Also, I added a generic flash message for failing to create the image

File: test/controllers/image/images_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    assert_response :unprocessable_entity
+    assert_select '.error', value: 'Invalid file format'
+    form_exists img[:image][:url]
+    assert_equal 'Failed to create image.', flash[:notice]
+  end
+
+  test 'should show image' do
+    assert_difference('Image.count') do
+      post images_path, params: { image: { url: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.png' } }
```
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: Since we're creating an image just for test setup, and we're not trying to test the `create` action in the test for the `show` action. We can just use `image = Image.create!(url: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.png')` instead of
```
assert_difference('Image.count') do
      post images_path, params: { image: { url: 'https://homepages.cae.wisc.edu/~ece533/images/airplane.png' } }
end
```

File: test/controllers/image/images_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+    image = Image.last
+    get image_path(image)
+
+    assert_response :ok
+    assert_select 'img', src: image.url
+  end
+
+  test 'should not show image' do
+    get image_path(-1)
```
Comment (jmangel) [inline test/controllers/image/images_controller_test.rb]: 👍
